lib/utils.js

```diff
@@ -296,6 +315,7 @@ module.exports = {
   isStream: isStream,
   isURLSearchParams: isURLSearchParams,
   isStandardBrowserEnv: isStandardBrowserEnv,
+  Clone: Clone,
```

Comment:
Nit-picky, but can you use a lower camel case name to match the other functions?

Reply:
I capitalized it since it's invoked using the `new` keyword which typically implies a "class," or a function that has constructor: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/new

However, if you still want me to lowercase it I can.. pls confirm.

Reply:
Ah gotcha, I didn't catch that change. This is fine as is.

Reply:
okidoki, thank you

Reply:
@heisian I don't see why `Clone` needs to be a class constructor. Why can't it be a regular function that returns a new object instead?

Reply:
@nickuraltsev The functionality of `new` is the key here. It guarantees that the object being assigned to `this[prop]` is a completely new object. Using `Object.assign({{}, obj[prop])` doesn't work here, nor does `Object.create(obj[prop])`, you can try it yourself.

If you have any other ideas, I'm open, but this seems to be the simplest way to recursively create a perfect clone of an object. Again, the `new` keyword and its inherent functionality are what enables this to work at all.

Reply:
If you want to avoid a class constructor which really isn't even a class in the ES6 sense but a _constructor function_ which has been readily available and widely used since _Javascript 1.0_, be my guest, but please propose an alternative solution that works here.

Reply:
Yes, that is what your `utils.merge` function does, where it was written: `var result = {}`. Clearly, however, this is not working to "separate" a new instance `defaults` object from the global axios `defaults` object.

..which is the very root of this issue. So if the current behaviour has a bug, this PR is intended to fix it.

Having the `defaults` of a separate instance of axios be able to modify all other instances of axios in a single script completely negates the point of having separate instances to begin with, IMO.

Do you not agree?

Reply:
I don't think this issue has anything to do with the fact that `merge` creates plain objects. The root cause is that `merge` does not always create a new object. For example, if `headers` exists in the source object, but does not exist in the target object, `merge` will take the source `headers` as is (without cloning it). If `headers` exists in both objects, `merge` will create a new object. 

Unfortunately, simply cloning the config object will fix some of the issues but will also introduce new ones. That is discussed here: https://github.com/axios/axios/pull/533#issuecomment-262427146

It's also discussed here: https://github.com/axios/axios/issues/370